src/send-linux.c

```diff
@@ -0,0 +1,105 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define _GNU_SOURCE
+#include <sys/socket.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <netpacket/packet.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/time.h>
+#include <linux/netlink.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+
+#include "../lib/includes.h"
+#include "./send.h"
+#include "./send-linux.h"
+
+#include <netpacket/packet.h>
+
+
+int send_run_init(sock_t s)
+{
+	// Get the actual socket
+	int sock = s.sock;
+	// get source interface index
+	struct ifreq if_idx;
+	memset(&if_idx, 0, sizeof(struct ifreq));
+	if (strlen(zconf.iface) >= IFNAMSIZ) {
+		log_error("send", "device interface name (%s) too long\n",
+			  zconf.iface);
+		return EXIT_FAILURE;
+	}
+	strncpy(if_idx.ifr_name, zconf.iface, IFNAMSIZ - 1);
+	if (ioctl(sock, SIOCGIFINDEX, &if_idx) < 0) {
+		log_error("send", "%s", "SIOCGIFINDEX");
+		return EXIT_FAILURE;
+	}
+	int ifindex = if_idx.ifr_ifindex;
+
+	// destination address for the socket
+	memset((void *)&sockaddr, 0, sizeof(struct sockaddr_ll));
+	sockaddr.sll_ifindex = ifindex;
+	sockaddr.sll_halen = ETH_ALEN;
+	if (zconf.send_ip_pkts) {
+		sockaddr.sll_protocol = htons(ETHERTYPE_IP);
+	}
+	memcpy(sockaddr.sll_addr, zconf.gw_mac, ETH_ALEN);
+	return EXIT_SUCCESS;
+}
+
+int send_batch(sock_t sock, batch_t* batch, int retries) {
+	struct mmsghdr msgvec [batch->capacity]; // Array of multiple msg header structures
+	struct msghdr msgs[batch->capacity];
+	struct iovec iovs[batch->capacity];
+
+	for (int i = 0; i < batch->len; ++i) {
+		struct iovec *iov = &iovs[i];
+	    	iov->iov_base = ((void *)batch->packets) + (i * MAX_PACKET_SIZE);
+	       	iov->iov_len = batch->lens[i];
+		struct msghdr *msg = &msgs[i];
+		memset(msg, 0, sizeof(struct msghdr));
+		// based on https://github.com/torvalds/linux/blob/master/net/socket.c#L2180
+		msg->msg_name = (struct sockaddr *)&sockaddr;
+		msg->msg_namelen = sizeof(struct sockaddr_ll);
+		msg->msg_iov = iov;
+		msg->msg_iovlen = 1;
+		msgvec[i].msg_hdr = *msg;
+		msgvec[i].msg_len = batch->lens[i];
+	}
+	int rv = 0;
+	for (int i = 0; i < retries; i++) {
+		// according to manpages
+		// On success, sendmmsg() returns the number of messages sent from msgvec; if this is less than vlen, the
+		//       caller can retry with a further sendmmsg() call to send the remaining messages.
+		// On error, -1 is returned, and errno is set to indicate the error.
+		rv = sendmmsg(sock.sock, &msgvec, batch->len, 0);
+		if (rv < 0) {
+			// only retry if all messages failed to send
+			perror("error in sendmmsg");
+		} else {
+			break;
```

Comment:
I'm not sure if I understand the logic here offhand. If there's a major error we retry, but if there's only the ability to do a partial send, then the function returns. The comment below (when the function is called) says:

> // whether batch succeeds or fails, this was the only attempt. Any re-tries are handled within batch

but I don't think this is true because if a fraction of messages aren't sent, then they aren't retried.

Reply:
That's true, and if this is the error-handling we want I should correct the comment. 
From my reading of the `man sendmmsg`, the caller _could_ attempt to re-send the packets starting with the first that failed, but I figured this wasn't really necessary. The user will get an error message on why the batch failed and can retry.

Do you think it'd be better to:

- Batch succeeds -> return
- Batch partially fails, -> retry the failed packets unless you're out of retries, then return
- Batch completely fails -> retry the entire batch, unless you're out of retries, then return

?

Reply:
If we didn't have retry behavior and were to just crash on send fail, it'd be different, but I think in our case, we're making the promise to the user _to retry_ based on having the command line option. I think that we should retry for the packets that did not successfully send. Generally, we shouldn't resend packets that were sent successfully since this creates load on the destination system.